[CDAP-13213][Cloud] Adds presentation layer for Create Profile view

ajainarayanan opened the pull request:
- Adds the initial version of Profile create view for cloud.
- Its a simple form with an initial set of values that UI can submit to backend.

#### Widgets:

- Adds the initial version of a widget system in react. I have tried to keep the design simple and something we are familiar with.
- #### From UI development,
  - Every widget will get three props,
    - `value`: The value of the widget
    - `onChange`: The onChange callback to be called when the widget value changes.
    - `widgetProps`: Extra props needed for the widget from the spec (like options, size etc.,)
  - Every widget is wrapped in a `StateWrapper` which will allow individual widgets to be stateless components
  - The `StateWrapper` is supposed to handle the state change and call the `onChange` from the parent for udpates.
  - `AbstractWidgetFactory` is similar to our `widget-factory`. It has a simpler proxy to return `input[type="text"]` if user specifies an unknown widget.

- #### From Profile/spec development,
  - The spec json right now has only `configuration-groups` similar to the widget json spec we have.
     
JIRA: https://issues.cask.co/browse/CDAP-13213

elfenheart commented on cdap-ui/app/cdap/components/Cloud/Profiles/CreateView/index.js:
```diff
@@ -0,0 +1,211 @@
+                      />
+                    }
+                  </Col>
+                </FormGroup>
```
Would it be better to extract this as it's own component?

ajainarayanan replied:
I am not sure if it will be of much use. This function is just iterating over the groups and rendering the widgets. May be when a group becomes a little bit more sophisticated like collapsable groups, some highlights, metadata (#of properties), subtitle etc.,

elfenheart commented on cdap-ui/app/cdap/components/AbstractWidget/AbstractWidgetFactory.js:
```diff
@@ -0,0 +1,38 @@
+  textarea: TextArea,
+  number: NumberTextbox,
+  select: SelectWidget,
+  'memory-dropdown': MemorySelectWidget,
```
isn't this just a normal dropdown?

ajainarayanan replied:
So right now the MemorySelectWidget and MemoryTextBox are simpler inputs with the extra label on the right. Simpler to extract as component if we are going to reuse.

tonybach commented on cdap-ui/app/cdap/components/AbstractWidget/NumberTextbox/index.js:
```diff
@@ -0,0 +1,40 @@
+require('./NumberTextbox.scss');
+
+export default function NumberTextbox({...props}) {
+  let {onChange, value, widgetProps = {} } = props;
```
nit: remove space before closing brace?

tonybach commented on cdap-ui/app/cdap/components/Cloud/Profiles/CreateView/index.js:
```diff
@@ -0,0 +1,211 @@
+    profileDescription: ''
+  };
+
+  parseSpecAndgetInitialState = () => {
```
nit: `parseSpecAndGetInitialState`?

tonybach commented on cdap-ui/app/cdap/components/SelectWithOptions/index.js:
```diff
@@ -45,17 +45,19 @@ export default function SelectWithOptions({className, value, onChange, options})
   value: ''
 };
-SelectWithOptions.propTypes = {
+export const SelectWithOptionsPropTypes = {
```
You will be using this constant somewhere?

tonybach commented on cdap-ui/app/cdap/components/AbstractWidget/index.js:
```diff
@@ -0,0 +1,51 @@
+
+export const WIDGET_PROPTYPES = {
+  widgetProps: PropTypes.object,
+  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
```
should this include `PropTypes.array` too, since that's also a possible value for the `value` prop of `AbstractWidget`?

tonybach commented on cdap-ui/app/cdap/components/Cloud/Profiles/CreateView/index.js:
```diff
@@ -0,0 +1,211 @@
+                    {
+                      <AbstractWidget
+                        type={property['widget-type']}
+                        value={""}
```
nit: maybe don't need the curly braces?

tonybach commented on cdap-ui/app/cdap/components/Cloud/Profiles/CreateView/index.js:
```diff
@@ -0,0 +1,211 @@
+          <Button
+            color="primary"
+            onClick={this.createProfile}
+            disabled={!this.state.profileName.length && !this.state.profileDescription.length}
```
I think this should be `||` instead of `&&`. Right now if I fill in the description but not the name, the button is still enabled.

tonybach commented:
You might also want to double check the height of the container. Right now the footer is overlapping with the content:

![create-profile-view](https://user-images.githubusercontent.com/6516002/37799533-37119dfc-2ddd-11e8-84d0-a035b264d855.gif)

tonybach commented:
Code LGTM 👍

tonybach committed:
[CDAP-13213][Cloud] Adds presentation layer for Create Profile view

  - Adds the widget json spec in react
  - Adds initial version of profile create view

ajainarayanan commented:
Demoed to @bdmogal . Merging.
